src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs

```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
             throw CreateExceptionForOSStatus(osStatus);
         }
 
-        internal static SafeTemporaryKeychainHandle CreateTemporaryKeychain()
+        internal static unsafe SafeTemporaryKeychainHandle CreateTemporaryKeychain()
         {
+            const int randomSize = 256;
             string tmpKeychainPath = Path.Combine(
                 Path.GetTempPath(),
                 Guid.NewGuid().ToString("N") + ".keychain");
 
             // Use a distinct GUID so that if a keychain is abandoned it isn't recoverable.
-            string tmpKeychainPassphrase = Guid.NewGuid().ToString("N");
+            // We use stack to minimize lingering
+            Span<byte> random = stackalloc byte[randomSize];
+            RandomNumberGenerator.Fill(random);
+
+            // Create hex-like UTF8 string.
+            Span<byte> utf8Passphrase =  stackalloc byte[randomSize * 2];
+            for (int i = 0; i< random.Length; i++)
+            {
+                // Instead of true hexadecimal, we simply take lower and upper 4 bits and we offset them from ASCII 'A'
+                // to get printable form. We dont use managed string to avoid lingering copies.
+                utf8Passphrase[i*2] = (byte)((random[i] & 0x0F) + 65);
+                utf8Passphrase[i*2 + 1 ] = (byte)((random[i] >> 4) & 0x0F + 65);
+            }
 
-            byte[] utf8Passphrase = System.Text.Encoding.UTF8.GetBytes(tmpKeychainPassphrase);
+            // clear the binary bits.
+            random.Fill(0);
```

Comment:
```suggestion
            CryptographicOperations.ZeroMemory(random);
```

Reply:
updated. Is that better for some reason then the Clear() @vcsjones was suggesting?

Reply:
Currently? No.  C compilers have always reserved the right (much to data security people's chagrin) to ignore memory writes that they can prove are never read again; so then calls like ZeroMemory get born to say "this one is important to clear".

If C#/.NET ever starts making those same optimizations, ZeroMemory will be immune.  We don't really expect to do that, but it's there just in case.  (It's implemented as a non-inlinable call to Clear :smile:)

Reply:
that make sense. thanks for the explanation @bartonjs